language-plutus-core/src/Language/PlutusCore/Quote.hs

```diff
@@ -44,7 +52,31 @@ emptyFreshState = Unique 0
 -- fresh-name generation, and parsing.
 newtype QuoteT m a = QuoteT { unQuoteT :: StateT FreshState m a }
     -- the MonadState constraint is handy, but it's useless outside since we don't export the state type
-    deriving (Functor, Applicative, Monad, MonadTrans, MM.MFunctor, MonadState FreshState)
+    deriving (Functor, Applicative, Monad, MonadTrans, MM.MFunctor, MonadState FreshState, MonadError e, MonadReader r)
+
+-- | A monad that allows lifting of quoted expressions.
+class Monad m => MonadQuote m where
+    liftQuote :: Quote a -> m a
+    -- This means we don't have to implement it when we're writing an instance for a MonadTrans monad. We can't just
+    -- add an instance declaration for that, because it overlaps with the base instance.
+    default liftQuote :: (MonadQuote n, MonadTrans t, t n ~ m) => Quote a -> m a
+    liftQuote = lift . liftQuote
+
+instance (Monad m) => MonadQuote (QuoteT m) where
+    liftQuote = MM.hoist (pure . runIdentity)
+
+instance MonadQuote m => MonadQuote (StateT s m)
+instance MonadQuote m => MonadQuote (ExceptT e m)
+instance MonadQuote m => MonadQuote (ReaderT r m)
+
+-- | Map the errors in a 'MonadError' and 'MonadQuote' context according to the given function.
+convertErrors :: forall a b n o .
+  (MonadError b n, MonadQuote n)
+  => (a -> b)
+  -- this needs to have the forall so *we* can choose what to instantiate it to (i.e. ExceptT a Quote)
+  -> (forall m . (MonadError a m, MonadQuote m) => m o)
```

Comment:
Yes, you're right - in practice what the rank-2 type says is that this thing can be instantiated to `ExceptT a Quote`, so that might as well happen at the call site.